File: backend/src/services/payme.service.js
const crypto = require('crypto');
const db = require('../config/database');
const { getIO } = require('../config/socket');

const MERCHANT_ID = process.env.PAYME_MERCHANT_ID;
const SECRET = process.env.PAYME_IS_TEST === 'true'
  ? process.env.PAYME_TEST_SECRET_KEY
  : process.env.PAYME_SECRET_KEY;

function verifyBasicAuth(authHeader) {
  if (!authHeader?.startsWith('Basic ')) return false;
  const decoded = Buffer.from(authHeader.slice(6), 'base64').toString();
  const [, password] = decoded.split(':');
  return password === SECRET;
}

async function handleWebhook(method, params, authHeader) {
  if (!verifyBasicAuth(authHeader)) {
    return { error: { code: -32504, message: 'Insufficient privilege' }, id: null };
  }

  switch (method) {
    case 'CheckPerformTransaction': return checkPerform(params);
    case 'CreateTransaction': return createTransaction(params);
    case 'PerformTransaction': return performTransaction(params);
    case 'CancelTransaction': return cancelTransaction(params);
    case 'CheckTransaction': return checkTransaction(params);
    default: return { error: { code: -32601, message: 'Method not found' }, id: null };
  }
}

async function checkPerform({ amount, account }) {
  const order = await db('orders').where({ id: account.order_id }).first();
  if (!order) return { error: { code: -31050, message: 'Order not found' } };
  if (order.payment_status === 'paid') return { error: { code: -31050, message: 'Already paid' } };
  if (Math.round(order.estimated_price * 100) !== amount) return { error: { code: -31001, message: 'Wrong amount' } };
  return { result: { allow: true } };
}

async function createTransaction({ id, time, amount, account }) {
  let payment = await db('payments').where({ external_id: id }).first();
  if (!payment) {
    [payment] = await db('payments').insert({
      order_id: account.order_id,
      amount: amount / 100,
      method: 'payme',
      status: 'processing',
      external_id: id,
    }).returning('*');
  }
  return { result: { create_time: new Date(time).getTime(), transaction: payment.id.toString(), state: 1 } };
}

async function performTransaction({ id }) {
  const payment = await db('payments').where({ external_id: id }).first();
  if (!payment) return { error: { code: -31003, message: 'Transaction not found' } };

  await db('payments').where({ id: payment.id }).update({ status: 'completed', completed_at: db.fn.now() });
  await db('orders').where({ id: payment.order_id }).update({ payment_status: 'paid' });

  const order = await db('orders').where({ id: payment.order_id }).first();
  const io = getIO();
  io.to(`passenger:${order.passenger_id}`).to(`driver:${order.driver_id}`).emit('payment_confirmed', {
    orderId: payment.order_id,
    amount: payment.amount,
    method: 'payme',
  });

  return { result: { transaction: payment.id.toString(), perform_time: Date.now(), state: 2 } };
}

async function cancelTransaction({ id, reason }) {
  const payment = await db('payments').where({ external_id: id }).first();
  if (!payment) return { error: { code: -31003, message: 'Transaction not found' } };
  await db('payments').where({ id: payment.id }).update({ status: 'failed' });
  return { result: { transaction: payment.id.toString(), cancel_time: Date.now(), state: -1 } };
}

async function checkTransaction({ id }) {
  const payment = await db('payments').where({ external_id: id }).first();
  if (!payment) return { error: { code: -31003, message: 'Transaction not found' } };
  const stateMap = { processing: 1, completed: 2, failed: -1 };
  return { result: { create_time: payment.created_at.getTime(), transaction: payment.id.toString(), state: stateMap[payment.status] || 1 } };
}

module.exports = { handleWebhook };

File: backend/app/services/payme_service.py
import base64
import asyncpg
from app.config import settings


def _verify_auth(auth_header: str | None) -> bool:
    if not auth_header or not auth_header.startswith("Basic "):
        return False
    decoded = base64.b64decode(auth_header[6:]).decode()
    _, password = decoded.split(":", 1) if ":" in decoded else ("", "")
    secret = settings.payme_test_secret_key if settings.payme_is_test else settings.payme_secret_key
    return password == secret


async def handle_webhook(method: str, params: dict, auth_header: str | None, conn: asyncpg.Connection) -> dict:
    if not _verify_auth(auth_header):
        return {"error": {"code": -32504, "message": "Insufficient privilege"}}

    match method:
        case "CheckPerformTransaction":
            return await _check_perform(params, conn)
        case "CreateTransaction":
            return await _create_transaction(params, conn)
        case "PerformTransaction":
            return await _perform_transaction(params, conn)
        case "CancelTransaction":
            return await _cancel_transaction(params, conn)
        case "CheckTransaction":
            return await _check_transaction(params, conn)
        case _:
            return {"error": {"code": -32601, "message": "Method not found"}}


async def _check_perform(params: dict, conn: asyncpg.Connection) -> dict:
    order_id = params.get("account", {}).get("order_id")
    order = await conn.fetchrow("SELECT * FROM orders WHERE id=$1", order_id)
    if not order:
        return {"error": {"code": -31050, "message": "Order not found"}}
    if order["payment_status"] == "paid":
        return {"error": {"code": -31050, "message": "Already paid"}}
    if round(float(order["estimated_price"]) * 100) != params.get("amount"):
        return {"error": {"code": -31001, "message": "Wrong amount"}}
    return {"result": {"allow": True}}


async def _create_transaction(params: dict, conn: asyncpg.Connection) -> dict:
    ext_id = params["id"]
    order_id = params.get("account", {}).get("order_id")
    amount = params.get("amount", 0)
    payment = await conn.fetchrow("SELECT * FROM payments WHERE external_id=$1", ext_id)
    if not payment:
        payment = await conn.fetchrow(
            "INSERT INTO payments(order_id,amount,method,status,external_id) VALUES($1,$2,'payme','processing',$3) RETURNING *",
            order_id, amount / 100, ext_id,
        )
    import time
    return {"result": {"create_time": int(time.time() * 1000), "transaction": str(payment["id"]), "state": 1}}


async def _perform_transaction(params: dict, conn: asyncpg.Connection) -> dict:
    from app.socket_manager import sio
    import time
    payment = await conn.fetchrow("SELECT * FROM payments WHERE external_id=$1", params["id"])
    if not payment:
        return {"error": {"code": -31003, "message": "Transaction not found"}}
    await conn.execute("UPDATE payments SET status='completed',completed_at=NOW() WHERE id=$1", payment["id"])
    await conn.execute("UPDATE orders SET payment_status='paid' WHERE id=$1", payment["order_id"])
    order = await conn.fetchrow("SELECT * FROM orders WHERE id=$1", payment["order_id"])
    if order:
        data = {"order_id": payment["order_id"], "amount": float(payment["amount"]), "method": "payme"}
        await sio.emit("payment_confirmed", data, room=f"passenger:{order['passenger_id']}")
        if order["driver_id"]:
            await sio.emit("payment_confirmed", data, room=f"driver:{order['driver_id']}")
    return {"result": {"transaction": str(payment["id"]), "perform_time": int(time.time() * 1000), "state": 2}}


async def _cancel_transaction(params: dict, conn: asyncpg.Connection) -> dict:
    import time
    payment = await conn.fetchrow("SELECT * FROM payments WHERE external_id=$1", params["id"])
    if not payment:
        return {"error": {"code": -31003, "message": "Transaction not found"}}
    await conn.execute("UPDATE payments SET status='failed' WHERE id=$1", payment["id"])
    return {"result": {"transaction": str(payment["id"]), "cancel_time": int(time.time() * 1000), "state": -1}}


async def _check_transaction(params: dict, conn: asyncpg.Connection) -> dict:
    payment = await conn.fetchrow("SELECT * FROM payments WHERE external_id=$1", params["id"])
    if not payment:
        return {"error": {"code": -31003, "message": "Transaction not found"}}
    state_map = {"processing": 1, "completed": 2, "failed": -1}
    return {"result": {"transaction": str(payment["id"]), "state": state_map.get(payment["status"], 1)}}
